### DISCUSSION TITLE
Adding a custom AirSim (Python) -> Unreal Engine (C++) function

### DISCUSSION THREAD

**Discussion by TomAvrech** [microsoft/AirSim#3337] [Support Q&A] [answered]
..\Unreal Projects\blocks\Plugins\AirSim\Source\WorldSimApi.h  has:  virtual void simFlushPersistentMarkers() override; .
Let's say I want to add my function to WorldSimApi.h : virtual void myFunc() override;
I first do it, and then implement the function in WorldSimApi.cpp : void WorldSimApi::myFunc() {...} ,
and I also add an abstract function to:
..\Unreal Projects\blocks\Plugins\AirSim\Source\AirLib\include\api\WorldSimApiBase.hpp : virtual void myFunc() = 0;
(It compiles)

Then, I want to call myFunc() from the same Pythonic place simFlushPersistentMarkers() is being called:
self.client.myFunc() which calls:
..\AirSim\PythonClient\airsim\client.py's function (I added it):
    def myFunc(self):
        self.client.call('myFunc')

But, when I do, I get the below error:
<img width="658" alt="Screenshot 2021-01-25 035426" src="https://user-images.githubusercontent.com/19329110/105652062-1ccf9f80-5ec1-11eb-9515-aac9a8c26f2c.png">
msgpackrpc.error.RPCError: rpclib: server could not find function 'myFunc' with argument count 0.

How do I add my C++ myFunc() to the "API", and do the other required actions to make it work ?

**Comment by ahmed-elsaharti** [comment, marked as answer]
I recommend having a look at PRs/commits where APIs (or any changes similar to the ones you want to make) were added in general to know what needs to be changed. For example, [here's](https://github.com/microsoft/AirSim/commit/0de1110d652451887adb049a5ec686f6cc201ef2) the commit where simFlushPersistentMarkers() was added. Also, [here's](https://github.com/microsoft/AirSim/pull/3242) a draft PR where I attempted to add an API that returns a custom struct through a custom function.

Hope this helps

**Reply by TomAvrech** [reply]
I have tried the simFlushPersistentMarkers() commit you posted, added the "same" function as simFlushPersistentMarkers() is, but with a different name (myFunc()), recompiled, and still got the same error as above from Python.

**Reply by ahmed-elsaharti** [reply]
Are you recompiling within Visual Studio or do you re-build using the build.cmd script?

**Reply by TomAvrech** [reply]
Compiling with VS alone didn't work.
So, I tried to build with build.cmd (it worked), then when I try to run the project from VS, it builds in VS too, runs, but still the problem remains in Python.

**Reply by ahmed-elsaharti** [reply]
And are your changes being made in the main AirSim.sln solution or in the Blocks.sln solution under Unreal/Environments/Blocks? 

The reason I'm asking is that I used to face similar issues when trying to add a new bit of code and testing it within the blocks environment. The workaround I used was to make changes within Blocks.sln, run the update_to_git script, build the main directory through build.cmd, run the update_from_git script and then run the Blocks.sln solution in VS.

**Reply by ahmed-elsaharti** [reply]
Also, I just tested copying the changes in the commit mentioned to add a myFunc() function (which in my case was set to reset the simulation) and it worked perfectly.

![build](https://user-images.githubusercontent.com/49626509/106219374-d61ec580-619e-11eb-8045-750286a79491.gif)

I think the issue you were facing was that the updates you were making weren't being added to the Blocks environment you're running which is why when the python client tried to call myFunc it got a response saying that the function could not be found on the server's end

**Reply by TomAvrech** [reply]
Changing the code in /AirSim/ , then using /AirSim/clean.cmd and /AirSim/build.cmd , then copying \AirSim\Unreal\Plugins\AirSim into my project's \Unreal Projects\blocks\Plugins solved.

Thanks, Ahmed! (I deleted some off-topic replies, maybe you should too)
